cpp/collective/rendezvous/include/utils.hpp

```diff
@@ -19,18 +19,273 @@ limitations under the License. */
+
+    const std::string &backtrace() const { return backtrace_; }
+
+    /// Returns the complete error message, including the source location.
+    /// The returned pointer is invalidated if you call add_context() on
+    /// this object.
+    const char *what() const noexcept override { return what_.c_str(); }
+
+    const void *caller() const noexcept { return caller_; }
+
+    /// Returns only the error message string, without source location.
+    /// The returned pointer is invalidated if you call add_context() on
+    /// this object.
+    const char *what_without_backtrace() const noexcept {
+        return what_without_backtrace_.c_str();
+    }
+
+private:
+    void refresh_what() {
+        what_ = compute_what(/*include_backtrace*/ true);
+        what_without_backtrace_ = compute_what(/*include_backtrace*/ false);
+    }
+    std::string compute_what(bool include_backtrace) const {
+        std::ostringstream oss;
+
+        oss << msg_;
+
+        if (context_.size() == 1) {
+            // Fold error and context in one line
+            oss << " (" << context_[0] << ")";
+        } else {
+            for (const auto &c : context_) {
+                oss << "\n  " << c;
+            }
+        }
+
+        if (include_backtrace) {
+            oss << "\n" << backtrace_;
+        }
+
+        return oss.str();
+    }
+};
+Error::Error(std::string msg, std::string backtrace, const void *caller)
+    : msg_(std::move(msg)), backtrace_(std::move(backtrace)), caller_(caller) {
+    refresh_what();
+}
+
+std::string charToString(const char *chars) {
+    std::string ret = chars;
+    return ret;
+}
+// Caffe2-style error message
+Error::Error(const char *file,
+             const uint32_t line,
+             const char *condition,
+             const std::string &msg,
+             const std::string &backtrace,
+             const void *caller)
+    : Error("[enforce fail at " + xoscar::StripBasename(file) + ":"
+                + std::to_string(line) + "] " + charToString(condition) + ". "
+                + msg,
+            backtrace,
+            caller) {}
+
+Error::Error(SourceLocation source_location, std::string msg)
+    : Error(std::move(msg),
+            "Exception raised from " + charToString(source_location.file)
+                + " (most recent call first):\n") {}
+
+void xoscarCheckFail(const char *func,
+                     const char *file,
+                     uint32_t line,
+                     const std::string &msg) {
+    throw ::xoscar::Error({func, file, line}, msg);
+}
+
+void xoscarCheckFail(const char *func,
+                     const char *file,
+                     uint32_t line,
+                     const char *msg) {
+    throw ::xoscar::Error({func, file, line}, msg);
+}
+
+#if defined(__GNUC__) || defined(__ICL) || defined(__clang__)
+#    define C10_LIKELY(expr) (__builtin_expect(static_cast<bool>(expr), 1))
+#    define C10_UNLIKELY(expr) (__builtin_expect(static_cast<bool>(expr), 0))
```

Comment:
It's better to rename the macro prefix. `C10_UNLIKELY` to `XOSCAR_UNLIKELY`.

Reply:
Done